### PR TITLE
fontconfig: fix subpixel rendering in GTK applications

### LINKED ISSUE #21429: subpixel rendering is broken in GTK programs
Hello,

RGB subpixel font rendering does not work as expected with GTK programs.

Enabling it in the Xfce settings panel gives a broken rendering across the whole desktop environment. Doing so in Plasma gives a very good rendering in KDE programs but a poor one within GTK programs.

While some other Linux systems fail at this too (Fedora, Mageia), most (including Debian, Manjaro, Solus, openSUSE...) do manage this properly and enabling RGB supixel rendering gives a very good rendering in all programs, both Qt and GTK.

* xuname:  
  *Void 5.4.35_1 x86_64 GenuineIntel uptodate rF*

### PR DESCRIPTION
**Pull request by andry-dev**
This PR hopefully closes #21429.

The first commit converts the freetype's ClearType patch to a build option; while it is not necessary for the fix, @ericonr suggested it and it seems helpful.

The second commit provides a `fontconfig` configuration file that sets `lcdfilter` to `lcddefault`, fixing the issue.
The config file provides a system-wide default without enforcing it, so the user/desktop environment can override the option in `~/.config/fontconfig/fonts.conf`. This doesn't break existing users: if they set the option anywhere then that is respected, if they didn't then subpixel rendering will work as intended.
Setting `lcdfilter` to `lcdlight` in `fonts.conf` is equivalent to the default "Harmony" engine, so it's the same as not having the patch in the first place.

## Why is this needed anyway?

A year ago I opened #13296, which enables ClearType in `freetype`. When testing it I had `lcdfilter` already set to `lcddefault` in my own `fonts.conf` and I didn't think about setting the option by default system-wide, so I never noticed any regression. Sorry for that.

I wrote the patch because without it I couldn't read text effortlessly, the font wasn't "right" for me. Major distros (Ubuntu, Debian and Fedora from the top of my head) have it enabled by default because, _generally speaking_, font rendering feels better with it.

### CODE REVIEW

File: srcpkgs/freetype/template [hunk 1]
```diff
@@ -12,6 +12,15 @@ license="GPL-2.0-or-later, FTL"
 homepage="https://www.freetype.org/"
 distfiles="${NONGNU_SITE}/${pkgname}/${pkgname}-${version}.tar.xz"
 checksum=1543d61025d2e6312e0a1c563652555f17378a204a61e99928c9fcef030a2d8b
+build_options="cleartype"
+desc_option_cleartype="Enable Microsoft's subpixel rendering (ClearType)"
+build_options_default="cleartype"
+
+pre_build() {
```
Comment (ahesford) [inline srcpkgs/freetype/template:19]: This shouldn't be in `pre_build`, it should be in `post_patch`.

File: srcpkgs/freetype/template [hunk 2]
```diff
@@ -12,6 +12,15 @@ license="GPL-2.0-or-later, FTL"
...
+	if [ yes = "$(vopt_if cleartype yes)" ]; then
+		vsed -i ${wrksrc}/include/freetype/config/ftoption.h -e 's@/\* #define FT_CONFIG_OPTION_SUBPIXEL_RENDERING \*/@#define FT_CONFIG_OPTION_SUBPIXEL_RENDERING@'
```
Comment (ahesford) [inline srcpkgs/freetype/template:21]: Drop `${wrksrc}/`, you're already in that directory in `post_patch`.

File: srcpkgs/freetype/template [hunk 3]
```diff
@@ -12,6 +12,15 @@ license="GPL-2.0-or-later, FTL"
 homepage="https://www.freetype.org/"
 distfiles="${NONGNU_SITE}/${pkgname}/${pkgname}-${version}.tar.xz"
 checksum=1543d61025d2e6312e0a1c563652555f17378a204a61e99928c9fcef030a2d8b
+build_options="cleartype"
+desc_option_cleartype="Enable Microsoft's subpixel rendering (ClearType)"
+build_options_default="cleartype"
```
Comment (ahesford) [inline srcpkgs/freetype/template:17]: Freetype itself says [emphasis added]
> Note that this feature is covered by several Microsoft patents and **should not be activated in any default build of the library**.

A build option is okay, but this should not be on by default.

File: srcpkgs/freetype/template [hunk 4]
```diff
@@ -12,6 +12,15 @@ license="GPL-2.0-or-later, FTL"
 homepage="https://www.freetype.org/"
 distfiles="${NONGNU_SITE}/${pkgname}/${pkgname}-${version}.tar.xz"
 checksum=1543d61025d2e6312e0a1c563652555f17378a204a61e99928c9fcef030a2d8b
+build_options="cleartype"
+desc_option_cleartype="Enable Microsoft's subpixel rendering (ClearType)"
```
Comment (ahesford) [inline srcpkgs/freetype/template:16]: Change to "Enable subpixel rendering similar to ClearType"

File: srcpkgs/freetype/template [hunk 5]
```diff
@@ -12,6 +12,15 @@ license="GPL-2.0-or-later, FTL"
...
+build_options_default="cleartype"
+
+pre_build() {
+	if [ yes = "$(vopt_if cleartype yes)" ]; then
```
Comment (ahesford) [inline srcpkgs/freetype/template:20]: `if [ "$build_option_cleartype" ]; then`

File: srcpkgs/fontconfig/template [hunk 6]
```diff
@@ -23,6 +23,8 @@ post_install() {
 	if [ "$CROSS_BUILD" ]; then
 		sed -i 's,\(Cflags: -I${includedir}\).*,\1,g' ${DESTDIR}/usr/lib/pkgconfig/fontconfig.pc
 	fi
+	vinstall ${FILESDIR}/11-lcdfilter-default.conf 644 usr/share/fontconfig/conf.avail
```
Comment (ahesford) [inline srcpkgs/fontconfig/template:26]: This file is already installed by fontconfig.

File: srcpkgs/fontconfig/template [hunk 7]
```diff
@@ -23,6 +23,8 @@ post_install() {
...
+	ln -s /usr/share/fontconfig/conf.avail/11-lcdfilter-default.conf ${DESTDIR}/etc/fonts/conf.d/
```
Comment (ahesford) [inline srcpkgs/fontconfig/template:27]: This is not the right way to handle this configuration snippet. In fact, fontconfig improperly installs several snippets in `/etc/fonts/conf.d`. The point of `/etc/fonts/conf.d` is to allow site-specific customization of fontconfig, but symlinks managed by the package will be recreated if the user deletes them and reinstalls or updates the package. Also, if the user removes the managed links, `xbps-pkgdb` will complain about broken symlinks when checking package integrity.

The right way to manage these symlinks is create an `INSTALL` script, run only for the initial installation and not an upgrade.

File: srcpkgs/freetype/template (hunk 3 above)
Comment (ericonr) [reply srcpkgs/freetype/template:17]: See #13296, it became available due to MS joining OIN.

Comment (ahesford) [reply srcpkgs/freetype/template:17]: Thanks for the clarification. I didn't look at the contents of the removed patch to see that we're already enabling this option. In that case, no harm in making it a default option.

**Comment by prez** [comment]
bump

**Comment by leahneukirchen** [comment]
I don't get why this PR is open. We have the subpixel rendering feature enabled by default, and a fresh installation of fontconfig also yields a `/etc/fonts/conf.d/11-lcdfilter-default.conf`.

**Comment by the-maldridge** [comment]
Closing as superfluous.  See @leahneukirchen's comment above.

**Pull request closed without merging**